lib/ruby_smb/server/share/provider/virtual_disk/virtual_pathname.rb

```diff
@@ -0,0 +1,193 @@
+
+            def <=>(other)
+              to_s <=> other.to_s
+            end
+
+            def exist?
+              !@stat.nil?
+            rescue Errno::ENOENT
+              false
+            end
+
+            def stat
+              raise Errno::ENOENT.new('No such file or directory') unless exist? && (@stat.file? || @stat.directory?)
+
+              @stat
+            end
+
+            def join(other)
+              # per the docs this Pathname#join doesn't touch the file system
+              # see: https://ruby-doc.org/stdlib-3.1.1/libdoc/pathname/rdoc/Pathname.html#class-Pathname-label-Core+methods
+              lookup_or_create(Pathname.new(to_s).join(other).to_s)
+            end
+
+            alias :+ :join
+            alias :/ :join
+
+            def to_s
+              @path
+            end
+
+            def absolute?
+              to_s.start_with?(SEPARATOR)
+            end
+
+            def relative?
+              !absolute?
+            end
+
+            def basename
+              lookup_or_create(self.class.basename(to_s))
+            end
+
+            def self.basename(*args)
+              File.basename(*args)
+            end
+
+            def dirname
+              lookup_or_create(self.class.dirname(to_s))
+            end
+
+            def self.dirname(*args)
+              File.dirname(*args)
+            end
+
+            def extname
+              File.extname(to_s)
+            end
+
+            def split
+              [dirname, basename]
+            end
+
+            alias :parent :dirname
+
+            def children(with_directory=true)
+              raise Errno::ENOTDIR.new("Not a directory @ dir_initialize - #{to_s}") unless directory?
+
+              @virtual_disk.each_value.select do |dirent|
+                next if dirent == self
+                next unless dirent.dirname == self
+
+                with_directory ? dirent : dirent.basename
+              end
+            end
+
+            def entries
+              children(false)
+            end
+
+            def cleanpath(consider_symlink=false)
+              lookup_or_create(self.class.cleanpath(to_s), stat: (exist? ? stat : nil))
+            end
+
+            def self.cleanpath(*args)
+              # per the docs this Pathname#cleanpath doesn't touch the file system
+              # see: https://ruby-doc.org/stdlib-3.1.1/libdoc/pathname/rdoc/Pathname.html#class-Pathname-label-Core+methods
+              Pathname.new(*args).cleanpath.to_s
+            end
+
+            private
+
+            # Check the virtual file system to see if the entry exists. Return it if it does, otherwise create a new
+            # entry representing a non-existent path.
+            #
+            # @param [String] path The path to lookup in the virtual file system. It will be normalized using #cleanpath.
+            # @return [Pathname] The path object representing the specified string.
+            def lookup_or_create(path, **kwargs)
```

Comment:
Because that's how the real Pathname works. If you take this example into consideration:

```
[2] pry(main)> p = Pathname.new('doesnotexist')
=> #<Pathname:doesnotexist>
[3] pry(main)> p.exist?
=> false
[4] pry(main)> p.stat
Errno::ENOENT: No such file or directory @ rb_file_s_stat - doesnotexist
from (pry):5:in `stat'
[5] pry(main)>
```

I tried to emulate it as close as I could using the VFS. So if you create a VirtualPathname object that represents a path that doesn't exist in the VFS we create a new object that doesn't exist and raise an exception when `#stat` is accessed. If the pathname does represent an entry on the VFS, its corresponding instance is returned. This makes sure that it works in the same way the standard `Disk` provider expects and handles requests to read files and list directories that don't exist in the same way.

Reply:
Oh ok, I see, thanks for the explanation.